### ISSUE TITLE
the GLPI Network registration key can not be saved and retruns a blank Page

### ISSUE THREAD

**Issue by edupulpillo** [glpi-project/glpi#8968]
**Describe the bug**
The GLPI Network registration key can not be saved and stored.
Clicking over the button "save" in /Home/Setup/Plugins/General/GLPI Network/Registration key, returns a blank page without displaying any error, and the content in the database of the field **glpinetwork_registration_key** in the table **glpi_configs** is empty. 

Version: GLPI 9.5.5 and 9.5.6 and 9.5.7
URL: /front/config.form.php?forcetab=GLPINetwork%241

![bug_key_registration](https://user-images.githubusercontent.com/7486607/115116525-877a0e00-9fcc-11eb-98c2-e70b4a022933.png)

**Comment by trasher** [comment]
A "blank page" is most of the time a fatal error, make sure to check all logs (including system ones)

**Comment by edupulpillo** [comment]
In Debug mode:
Uncaught Exception SodiumException: Key must be CRYPTO_AEAD_XCHACHA20POLY1305_KEYBYTES long in /home4/HOSTNAME/public_html/DOMAIN/glpi/vendor/paragonie/sodium_compat/src/Compat.php at line 897

More people observe the same problem:
https://forum.glpi-project.org/viewtopic.php?id=279454

**Comment by cedric-anne** [comment]
Hi,

As we need to be able to decrypt the value that is stored in DB, we use a reversible cryptographic algorithm provided by sodium library.
If you have this error, it means that the content of `config/glpicrypt.key` file is corrupted. Did you manually changed the content of this file ?
You should remove this file and use `glpi:security:change_key` CLI command to regenerate a valid encryption key.

Regards

**Comment by edupulpillo** [comment]
Hi, 
Thanks for following this case. No, i did not change manually the content of config/glpicrypt.key.
The problem is present with a fresh installation of GLPI.
So is not possible to set the GLPI Network registration key.

How can be accesed the GLPI console without SSH access, what is the Url?  (url_of_GLPI/..)

**Comment by trasher** [comment]
Hi,

> How can be accesed the GLPI console without SSH access, what is the Url? (url_of_GLPI/..)

The console is a cli tool, it cannot be accessed from an URL.

**Comment by edupulpillo** [comment]
Yes, but is writed in PHP, so mayeb very useful to be ablet o access it with admin credentials, from a URL. Keep in mind that shared hosting do not offer ssh access.

**Comment by edupulpillo** [comment]
May o ask why this topic is closed without a solution?

**Comment by trasher** [comment]
The solution is to use the `glpi:security:change_key` command.

Allowing that kind of feature from a URI is not something that will be done; mainly for security reasons.

There are for sure hostings that offers an ssh access, or maybe you can consider taking an already hosted GLPI solution. This is not a strict requirement, but in your case, there is no other solution; since your file has been corrupted. You can try to retrieve a non corrupted version in your backups maybe.

**Comment by cedric-anne** [comment]
Hi,

You can enventually install a fresh version of GLPI on another server (even on you local machine), in order to generate a new key file. Then you can get the generated file (`config/glpicrypt.key`) and copy it to your GLPI server.

You will have to retype your services passwords (mailcollector, ldap, ...), but as your current key file is corrupted, I guess stored passwords cannot be decoded.

Regards

**Comment by edupulpillo** [comment]
> `glpi:security:change_key` command

Thanks for your answer, but is not that my file get corrupted. A fresh installation always install a corrupted file, the soution should be for the next version to be sure that after the installation is not left a corrupted file.

**Comment by cedric-anne** [comment]
Hi,

Please provide all informations related to your system configuration. Encryption/decryption uses sodium PHP extension, or a polyfill if this extension is not available. Maybe you have an outdated sodium extension, but we need to know your system configuration to be able to investigate on this.

Regards

**Comment by edupulpillo** [comment]
The problem persist

**Comment by cedric-anne** [comment]
Hi,

Please provide all informations related to your system configuration. Encryption/decryption uses sodium PHP extension, or a polyfill if this extension is not available. Maybe you have an outdated sodium extension, but we need to know your system configuration to be able to investigate on this.

Regards

**Comment by edupulpillo** [comment]
The problem still is present, even installing GLPI from scratch. So the default installation does not allow to save a registration key to access to the GLPI Network

**Comment by edupulpillo** [comment]
The problem still is present in a fresh installation

**Comment by cedric-anne** [comment]
During installation process, have you some warnings errors (in UI or in logs) ?

Also, could you try to reproduce without having the sodium extension active ?

**Comment by filliperitzel** [comment]
No errors, only sodium is missing, i already tried with and without sodium.

**Comment by filliperitzel** [comment]
i have some warnings?

- For security reasons, SELinux mode should be Enforcing.For security reasons, SELinux mode should be Enforcing.
- Access to timezone database (mysql) is not allowed.Access to timezone database (mysql) is not allowed.
- sodium extension is not presentsodium extension is not present

I have a standard hosting and a VPS, I can't activate the market place on either one

**Comment by cconard96** [comment]
If your GLPI server cannot access the internet directly, you will have to configure the proxy settings in GLPI before the marketplace, network registration, and some other features will work.

**Comment by filliperitzel** [comment]
this erro is about internet acess?? 

Uncaught Exception SodiumException: Key must be CRYPTO_AEAD_XCHACHA20POLY1305_KEYBYTES long in /home4/rioeng69/bsq.rioeng.com.br/vendor/paragonie/sodium_compat/src/Compat.php at line 897

**Comment by cconard96** [comment]
No. That error specifically is probably from not having the sodium PHP extension installed and enabled or the libsodium package version is too low.

**Comment by filliperitzel** [comment]
I disabled sodium per cedric's guidance

on my VPS, sodium is active and the problem persist

**Comment by cconard96** [comment]
Then it is probably the second thing I mentioned. There is a library that GLPI uses in case the sodium extension is missing, but that library still relies on libsodium.

If you have a version that is too old, it may have issues.

**Comment by filliperitzel** [comment]
thnks, can u tell me wich version should i use?

**Comment by cconard96** [comment]
From my research, it looks like version 1.0.12 or higher is needed for GLPI.

On my system (Debian-based) `dpkg -l | grep 'sodium'` shows that I have "libsodium23:amd64                          1.0.18-1" installed.

The instructions for updating would vary depending on your OS.

**Comment by edupulpillo** [comment]
Hi. Unfortunately the problem still is present trying to save the GLPI Network registration key when click over button SAVE in /glpi/front/config.form.php:
Uncaught Exception SodiumException: secret key size should be SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES bytes in /home/public_html/glpi/inc/toolbox.class.php at line 275

**Comment by filliperitzel** [comment]
I'm building a centOS7 with all the specifications described on this page, I'll be back to tell you what works.

It's worth mentioning that I'm using PHP7.4 and all the tutorials that claim to work refer to PHP7.3 and its extensions.

**Comment by filliperitzel** [comment]
Dudes!! 
I finally got it! Below are the commands on centOS7. read it calmly because I made it for version 10 and version 9.5

[centOS 7 GLPI.txt](https://github.com/glpi-project/glpi/files/8105930/centOS.7.GLPI.txt)

**Comment by trasher** [comment]
Hi, note that `chcon` command will not persist the selinux flag on the files. After a relabel; the default context will be used again. Also, you should give selinux access only on `config` and `files` directories:

```
# semanage fcontext -d -t httpd_sys_rw_content_t '/path/to/glpi/*/(files|config)(/.*)?'
# restorecon -R -v /path/to/glpi/
```

**Comment by filliperitzel** [comment]
thanks, i will add these lines on my server. is the rest ok?

**Comment by filliperitzel** [comment]
to use "semanage" command, need to install "policycoreutils-python-utils" package

below are the procedures:

yum provides /usr/sbin/semanage

#check the version and edit the line below with the one that appears for you

yum install policycoreutils-python-2.5-34.el7.x86_64

#but when I run the Semanage command, it returns the following error.

ValueError: File context for /path/to/glpi/(files|config)(/.*)? is not defined

**Comment by trasher** [comment]
Sorry for the `-d` instead of `-a`; I've copied thee wrong command... Remember you have to run `restorecon` command to apply new flag.

About the initial problem, the `libsodium` installed on your system is outdated; you cannot use the php extension. Anyways, that is not really important because GLPI provide an alternative PHP library. The main difference is speed; but that does not really matters for GLPI usage currently.

#10750 will provide a clear error message about that.

**Comment by cconard96** [comment]
`libsodium` 1.0.12 or higher is currently required for both the PHP extension and the alternative library since both use it and at least one of the features used by GLPI was added in that version of the library. The alternative PHP library only replaces the functionality of the `php-sodium` extension and doesn't change or replace the functionality of the underlying library.

It sounds like the solution here is to update the library package version. I'll close this issue as the issue doesn't seem to be on the GLPI side. If you need other help with GLPI (non-bugs), you can get help on the [community forum](https://forum.glpi-project.org/index.php).

**Comment by edupulpillo** [comment]
It should not be closed.
We are using phpinfo shows a ibsodium headers version 1.0.16 and libsodium library version 1.0.16, that is higher than 1.0.12 and the problem persist.

**Comment by cedric-anne** [comment]
@edupulpillo @filliperitzel
I could help if you could provide a Dockerfile that can be used to reproduce your issue easilly ? There is no need to provide the SQL server in this docker container, the PHP environment is sufficient. I have no experience with CentOS and SeLinux, and unfortunately I don't have time to build a complete environment.

**Comment by filliperitzel** [comment]
from what I've seen so far, the problem is the hosting or system update.  I suggest using Hostiger, they don't have a glpi installed, but it allows you to enable glpi extensions.  mine is working 100% with them.  now every time i updated or cloned glpi it presented the error with the "corrupted" key.  But then, those of you who are wanting the marketplace were deceived, there is nothing good there, and what is good is paid and VERY EXPENSIVE.  I spent 1 month on this odyssey just to use GAPP and I have the terrible feeling that soon it will be paid too.  I've already started php to make my own plugins.  It's disappointing to see people trying to monetize on something free... but that's what free knowledge is, there's always a greedy starving person who charges for what they got for free

**Comment by cedric-anne** [comment]
> I think should exist a function to check the consistency of the stored registration key, and if an error occurs, or the key appears corrupted, then ask it again to the user.

In GLPI 10, some checks have been added, for instance in #9729.

Problem with current issue is that we do not reproduce it, so it is quiet hard to fix in these conditions.

**Comment by orthagh** [comment]
> But then, those of you who are wanting the marketplace were deceived, there is nothing good there, and what is good is paid and VERY EXPENSIVE. I spent 1 month on this odyssey just to use GAPP and I have the terrible feeling that soon it will be paid too. I've already started php to make my own plugins. It's disappointing to see people trying to monetize on something free... but that's what free knowledge is, there's always a greedy starving person who charges for what they got for free

Is this was really necessary ?

I mean, you currently getting **free** support on the topic. And 3 developers of GLPI are trying to help you.
The issue, which has not been ignored by us, is not replicated on our side, seems to be related to a specific environment, and we require more information to found a bypass fix on the core.

I suggest, maybe, you can keep a cool tone, no ?

**Comment by edupulpillo** [comment]
Hi [orthagh](https://github.com/orthagh) completely agree about your mentioning "you can keep a cool tone, no ?"
Opensource projects are great, and means sharing with the community, so even your message was not for me, i want to highlight: thank you for your product and for your support, and for your sharing in my case!
At the same time, please kindly take my comments and bug descriptions about bugs as my small contributions to the project. Because this is really my goal :-)

Returning to the bug:
Unfortunately i did not start to use your product only because exist this bug about the modules installation, that is very convenient. 
Following the thread, it seems we are not sync about the topic. So trying to summarize to help you to understand the problem:
- I have linux installed with PHP and using phpinfo shows a ibsodium headers version 1.0.16 and libsodium library version 1.0.16
- I have uninstalled and deleted all the files of GLPI, and installed from scratch even in different path the product, and the problem persist.

So, seems that the GLPI installation wizzard publised in Softaculous has a problem, or that is not recognized well libsodium. Maybe you can drive us hopw to check deeper the problem? wich file and source code line is involved around, to have a look?
Is there any procedure to provide you logs without a command line requirement?
cheers

**Comment by orthagh** [comment]
> At the same time, please kindly take my comments and bug descriptions about bugs as my small contributions to the project. Because this is really my goal :-)

Fine with me and thank you for the clarification.

> Softaculous

This is the first mention to this third party tool.
We cannot say it the culprit here, but it worth trying without it is possible
You can get the last build of glpi here (take the master archive) : https://nightly.glpi-project.org/glpi/
and install it on your server to check if the issue is still here

**Comment by edupulpillo** [comment]
Softaculous is one of the most famous automated deployment tool for shared hosting, that only execute the installation wizzard of multiple opensource products. So we have the problem in the wizzard for sure, i will check the version 10 as [cedric-anne](https://github.com/cedric-anne) suggest.

**Comment by cedric-anne** [comment]
I tried to find Softaculous scripts source, but I did not find any public repository related to them, so I cannot tell if their script if buggy or if the issue is related to tour system configuration.

As I already said, having a docker image on which problem is reproducible, on current [GLPI master nightly build](https://nightly.glpi-project.org/glpi/) could help us to fix this issue.

**Comment by edupulpillo** [comment]
Hi!. Softaculous and ansible puppet, cheff  and all the another deployment tools, does not touch the source code or files of the opensource platform, so this is not frecuently a reason about source of bugs.
Also is not good to make mandatory that only a docker version must be used to be able to deploy GLPI, under my point of view. In the configuration of my system, as i mentioned libsodium is available, and i do not find any problem on it.
The problem seems was in a file of the installed of GLPI, or in the source code of GLPI that do not provide information warning the user if can not reach the library libsodium.so , so in any case GLPI should not return an "empty page" and should display an error page, or whatever feedback, instead an empty page.

Im waiting you release the new version that i see you have introduced some new feedback in the case that an error occurs. Lets cross the fingers.

**Comment by edupulpillo** [comment]
Th eproblem is still present, and there is no way to fix it.
Using a fresh installation, the problem persist

**Comment by cedric-anne** [comment]
When registration key cannot be generated, an error message should be displayed in UI or console during installation, and messages should probably be put in logs.
These messages could help to identify the problem.

**Comment by edupulpillo** [comment]
How to get the logs using the web interface ? where are these logs located / what is the path of the logs in the filesystem needed to check ?

**Comment by cedric-anne** [comment]
> How to get the logs using the web interface ?

On UI, you should have this error message:
![image](https://user-images.githubusercontent.com/33253653/171432358-714c062b-571b-4007-a52f-a94a554f6e07.png)

Logs cannot be accessed from UI, but they can be accessed on filesystem and are located in `files/_logs` directory of GLPI. The `php-errors.log` file should contains a trace for the error corresponding to encryption key generation failure.

**Comment by edupulpillo** [comment]
The problem was fixed in the version 10.0.3 of GLPI
